### PR TITLE
Change: Refactor failsafe for better readability

### PR DESCRIPTION
**Pull request by nickanderson**
Changelog: Title

ping @jimis

### CODE REVIEW

**Comment by nickanderson** [comment]
trigger build

File: libpromises/failsafe.cf [hunk 1]
```diff
@@ -30,25 +30,44 @@
...
-                          "cfe_internal_update",
-                          "cfe_internal_call_update"
-                        };
-}
-
 body agent control
 {
+      # We can't expect dns to be properly configured everywhere so we
+      # skip the dns check for bootstrap
       skipidentify => "true";
```
Comment (jimis) [inline libpromises/failsafe.cf]: `skipidentify` only affects classic protocol connections, which are no longer used by default.

File: libpromises/failsafe.cf [hunk 2]
```diff
@@ -30,25 +30,44 @@
...
 
       # Bootstrapping can't continue without keys
       abortclasses => { "no_ppkeys_ABORT" };
 }
 
-################################################################################
+###################################################################################################################################
```
Comment (jimis) [inline libpromises/failsafe.cf]: How many hash marks are too many? :-) I would say anything over the 78th column.

File: libpromises/failsafe.cf [hunk 3]
```diff
@@ -30,25 +30,44 @@
...
+
+bundle agent main
+{
+  meta:
+      "description"
+        stirng => "Perform bootstrap or failsafe recovery operations";
```
Comment (jimis) [inline libpromises/failsafe.cf]: s/stirng/string/

File: libpromises/failsafe.cf [hunk 4]
```diff
@@ -84,80 +109,116 @@ bundle agent cfe_internal_update
...
     windows::
 
-      "$(sys.workdir)\\bin-twin\\."
-      handle => "cfe_internal_bootstrap_update_files_sys_workdir_bin_twin_windows",
-      comment => "Make sure we maintain a clone of the binaries and libraries for updating",
-      copy_from => u_cp("$(sys.workdir)\\bin\\."),
-      depth_search => u_recurse("1");
-
+      # TODO: Remove the use of bin-twin
+      "$(sys.workdir)\\bin-twin\\." -> { "[[redmine:xxxx]]" }
```
Comment (jimis) [inline libpromises/failsafe.cf]: What is this syntax?

**Comment by jimis** [comment]
Nice refactoring, besides the minor comments. But does it include functionality change? I believe the checking if cf-serverd and cf-execd daemons are running, has changed, right? If so in what way? Maybe it should be separate.

File: libpromises/failsafe.cf (hunk 2 above)
Comment (nickanderson) [inline libpromises/failsafe.cf]: OK, thats fair, i was making them long enough to match the longest line

File: libpromises/failsafe.cf (hunk 4 above)
Comment (nickanderson) [inline libpromises/failsafe.cf]: oh thats some of my org stuff :)

File: libpromises/failsafe.cf (hunk 3 above)
Comment (nickanderson) [inline libpromises/failsafe.cf]: ack

File: libpromises/failsafe.cf (hunk 1 above)
Comment (nickanderson) [inline libpromises/failsafe.cf]: So we can redact the skipidentify all together then?

**Comment by nickanderson** [comment]
Functionally no change, only cleanup of classes to make it more simple (at least in my mind).

The functional changes as part of this re-factor are minimal.

1) No longer warning that execd was not started based on the fact that it wasn't started. Now it only warns if it fails to restart execd. Note the subtle difference between reporting a failure when it wasn't attempted, vs reporting a failure when we know there was a failure. This handles cases where execd is already running but failsafe triggers.

2) Previously we only checked for cf-serverd if we were the policy server and we had fetched policy successfully. Now we define the class for cf-serverd process not running on all hosts. Previously the class `start_server` was only used to to start cf-serverd on non systemd hosts (but remember that this class was only set on policy servers). Now we only run the command to start cf-serverd on the policy_server when cf_serverd_not_running and its not a systemd host. Which is basically the same as what we had before since we only defined the start_server class on policy hubs to begin with.

So the only net change should be not reporting failure to start execd when we didnt bother trying (which is I think what we want).

**Comment by nickanderson** [comment]
changed based on feedback and rebased.

**Comment by nickanderson** [comment]
ping @jimis any obections?

**Comment by jimis** [comment]
If your bootstrap and failsafe run tests work, then let me know and I'll merge! :-)

**Comment by jimis** [comment]
trigger build

**Comment by jimis** [comment]
hmmm, it didn't pass `make check` in Jenkins, did it pass in your local test?

```
00:05:07.764 Starting test: policy_test.c
00:05:07.764 ==================================================
00:05:07.764 test_failsafe: Starting test
00:05:07.766 FAIL: policy_test
00:05:07.860 ==================================================
```

**Comment by nickanderson** [comment]
Gah your right, I had some syntax errors in there.

I found and fixed 2 issues.

```
> $ cf-promises -cf ./libpromises/failsafe.cf  
```

**Comment by nickanderson** [comment]
Lets wait on this for https://github.com/cfengine/core/pull/2500/commits which fixes an issue for duplicate definition if you include the stdlib in def.json

**Comment by nickanderson** [comment]
superseded by #2510

**Pull request closed without merging**